rfd/0025-hsm.md

```diff
@@ -0,0 +1,378 @@
+This may be used to distribute a single CA private key to multiple HSMs over
+the network. However, it significantly weakens the security properties of using
+an HSM in the first place.
+
+#### Teleport CAs
+
+Each Teleport cluster has multiple CAs:
+- Host CA (SSH + TLS)
+- User CA (SSH + TLS)
+- JWT CA (TLS only)
+
+Private keys in each CA are stored as a list to enable CA rotation. For
+example, when rotating a Host CA, we shuffle the key fields of the existing CA
+object instead of creating a brand new backend object. The expectation of only
+1 CA of each type is deeply integrated in Teleport.
+
+### Options
+
+This is how the HSM integration could work, with increasing level of
+difficulty.
+
+#### Option 1: users manage one key, teleport uses it
+
+The easiest option - outsource CA key creation and replication to the user.
+Teleport receives only the following `teleport.yaml` options:
+
+- PKCS#11 module path
+- PIN for HSM device authentication
+- RSA key label within the HSM (key must already exist) for every CA type
+
+Teleport encodes this information in the private key field of the CA backend
+object (instead of the PEM-encoded raw private key) and uses its local PKCS#11
+device to perform CA signing operations.
+
+Teleport CA rotation could work with this approach by providing a second set of
+HSM key references in `teleport.yaml` and removing the original set after
+rotation is done.
+
+For HA Auth server setup, users would have to manually migrate the private keys
+to multiple HSM devices, one for each Auth server.
+
+Clearly, this option offers the worst UX but is easiest to implement.
+
+#### Option 2: users manage multiple keys, teleport uses them
+
+This is effectively Option 1 without migrating keys across HSM devices. Each
+Auth server would have a unique set of private keys.
+
+As a result, we need to update our CA storage schema and CA rotation to allow
+multiple active private keys for each CA type. We'll also need to audit all
+clients' usage of CA certificates during validation, to make sure they use them
+all and not only the first one.
+
+#### Option 3: teleport manages and uses multiple keys
+
+This option removes the burden of creating private keys from the user. On
+startup, the Auth server will create the key in the HSM as needed. As a
+side-effect, this gives us the flexibility of switching key algorithms in the
+future.
+
+We still need to support multiple private keys as in Option 2.
+
+#### Option 4: teleport manages and uses one key
+
+This option sticks with our current architecture of 1 active private key per CA
+type. Teleport Auth server will not only create the HSM key on first start, but
+also migrate an existing key from a different Auth server if there are any.
+The details of this auto-migration are not clear yet, we don't have a precedent
+of auth servers talking to each other directly.
+
+Compared to options 2 and 3, we won't need to update the CA rotation or clients
+much.
+
+#### awly's preference
```

Comment:
@awly I agree Option 3 is the most promising, l suggest we give it a go

Reply:
Ack, squashed all the options in the doc